src/Stratis.Bitcoin/P2P/Peer/NetworkPeerConnection.cs

```diff
@@ -173,40 +176,32 @@ public async Task ConnectAsync(IPEndPoint endPoint, CancellationToken cancellati
 
             try
             {
-                // This variable records any error occurring in the thread pool task's context.
-                Exception error = null;
-
-                await Task.Run(() =>
-                {
-                    try
-                    {
-                        this.tcpClient.ConnectAsync(endPoint.Address, endPoint.Port).Wait(cancellation);
-                    }
-                    catch (Exception e)
-                    {
-                        // Record the error occurring in the thread pool's context.
-                        error = e;
-                    }
-                }).ConfigureAwait(false);
-
-                // Throw the error within this error handling context.
-                if (error != null)
-                    throw error;
+                this.asyncProvider.Signals.Publish(new PeerConnectionAttempt(false, endPoint));
+                await this.tcpClient.ConnectAsync(endPoint.Address, endPoint.Port).WithCancellationAsync(cancellation).ConfigureAwait(false);
+                this.asyncProvider.Signals.Publish(new PeerConnected(false, endPoint));
 
                 this.stream = this.tcpClient.GetStream();
             }
             catch (OperationCanceledException)
             {
+                this.asyncProvider.Signals.Publish(new PeerConnectionAttemptFailed(false, endPoint, "Operation Canceled"));
                 this.logger.LogTrace("Connecting to '{0}' cancelled.", endPoint);
                 this.logger.LogTrace("(-)[CANCELLED]");
                 throw;
             }
-            catch (Exception e)
+            catch (SocketException ex)
             {
-                if (e is AggregateException) e = e.InnerException;
```

Comment:
I see this line got removed, is that intentional? Unless there is no possibility of an `AggregateException` getting thrown I think it is useful to know the inner exception.

Reply:
There is no chance because when you use await, the exception isn't kept in Task but thrown directly
Now I await the `this.tcpClient.ConnectAsync` so the exception handling is easier.